server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsResponse.java

```diff
@@ -19,102 +19,244 @@
+        return successfulResponses;
+    }
+
+    /**
+     * Returns a map of repository name to {@link ElasticsearchException} for each unsuccessful response.
+     */
+    public Map<String, ElasticsearchException> getFailedResponses() {
+        return failedResponses;
+    }
+
+    /**
+     * Returns true if there is a least one failed response.
+     */
+    public boolean isFailed() {
+        return failedResponses.isEmpty() == false;
     }
 
     @Override
-    public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
-        builder.startArray("snapshots");
-        for (SnapshotInfo snapshotInfo : snapshots) {
-            snapshotInfo.toXContent(builder, params);
+        builder.startArray("responses");
+
+        for (Map.Entry<String, List<SnapshotInfo>> snapshots : successfulResponses.entrySet()) {
+            builder.startObject();
+            builder.field("repository", snapshots.getKey());
+            builder.startArray("snapshots");
+            for (SnapshotInfo snapshot : snapshots.getValue()) {
+                snapshot.toXContent(builder, params);
+            }
+            builder.endArray();
+            builder.endObject();
         }
+
+        for (Map.Entry<String, ElasticsearchException> error : failedResponses.entrySet()) {
+            builder.startObject();
+            builder.field("repository", error.getKey());
+            ElasticsearchException.generateFailureXContent(builder, params, error.getValue(), true);
+            builder.endObject();
+        }
+
         builder.endArray();
         builder.endObject();
         return builder;
     }
 
-    public static GetSnapshotsResponse fromXContent(XContentParser parser) throws IOException {
-        return GET_SNAPSHOT_PARSER.parse(parser, null);
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        if (out.getVersion().onOrAfter(GetSnapshotsRequest.MULTIPLE_REPOSITORIES_SUPPORT_ADDED)) {
+            out.writeVInt(successfulResponses.size());
+            for (Map.Entry<String, List<SnapshotInfo>> snapshots : successfulResponses.entrySet()) {
+                out.writeString(snapshots.getKey());
+                out.writeVInt(snapshots.getValue().size());
+                for (SnapshotInfo snapshotInfo : snapshots.getValue()) {
+                    snapshotInfo.writeTo(out);
+                }
+            }
+            out.writeVInt(failedResponses.size());
+            for (Map.Entry<String, ElasticsearchException> error : failedResponses.entrySet()) {
+                out.writeString(error.getKey());
+                out.writeException(error.getValue());
+            }
+        } else {
+            if (successfulResponses.size() + failedResponses.size() != 1) {
```

Comment:
As I mentioned above,this is kind of weird to me. How would an older node request multiple repositories in the first place? It seems like we don't have to worry about this, but instead have to throw on newer nodes when master is too old.